File: app/src/main/java/com/meetferrytan/mvpdaggerstore/util/format/BufferedSourceConverterFactory.java
package com.meetferrytan.mvpdaggerstore.util.format;

import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import okio.BufferedSource;
import retrofit2.Converter;
import retrofit2.Retrofit;

public final class BufferedSourceConverterFactory extends Converter.Factory {
    @Override
    public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations,
                                                            Retrofit retrofit) {
        if (!BufferedSource.class.equals(type)) {
            return null;
        }
        return (Converter<ResponseBody, BufferedSource>) ResponseBody::source;
    }
}

File: app/src/main/java/com/meetferrytan/mvpdaggerstore/util/format/BufferedSourceConverterFactory.kt
package com.meetferrytan.mvpdaggerstore.util.format

import okhttp3.ResponseBody
import okio.BufferedSource
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

class BufferedSourceConverterFactory : Converter.Factory() {
    override fun responseBodyConverter(type: Type?, annotations: Array<Annotation>?,
                                       retrofit: Retrofit?): Converter<ResponseBody, *>? {
        return if (BufferedSource::class.java != type) {
            null
        } else Converter<ResponseBody, BufferedSource> { it.source() }
    }
}